File: 프로그래머스/Programmers250137.java
import java.util.*;
class Solution {
    public int solution(int[] bandage, int health, int[][] attacks) {
        int answer = 0;
        int max = 0;
        int oH = health;
        Map<Integer,Integer> monGraph = new HashMap<>();
        for(int[] num : attacks){
            monGraph.put(num[0],num[1]);
            max  = Math.max(max,num[0]);
        }
        int bandageCombo = 0;
        for(int i=1;i<=max;i++){
            if(monGraph.containsKey(i)){
                health -= monGraph.get(i);
                bandageCombo = 0;
                if(health <= 0) return -1;
            }
            if(!monGraph.containsKey(i) && health <= oH){
                health += bandage[1];
                bandageCombo++;
                if(health > oH) health = oH;
            }
            if(!monGraph.containsKey(i) && bandageCombo == bandage[0]){
                health += bandage[2];
                bandageCombo =0;
                if(health > oH) health = oH;
            }
            System.out.println(i+"번째"+" 남은체력:"+health+" 콤보:"+bandageCombo);
        }
        
        answer = health;
        return answer;
    }
}

File: 프로그래머스/Programmers250137.js

function solution(bandage, health, attacks) {
    let originHealth = health;
    let bonusTime = bandage[0];
    let healing = bandage[1];
    let bonusHealing = bandage[2];
    
    var answer = 0;
    let attackMap = new Map();
    let maxTime = 0;
    for(let i=0;i<attacks.length;i++){
        let key = attacks[i][0];
        let value = attacks[i][1];
        if(maxTime < key) maxTime = key;
        attackMap.set(key,value);
    }
    let continuityBandage = 0;
    for(let i=1;i<=maxTime;i++){
        continuityBandage++;
        if(attackMap.has(i)){
            continuityBandage = 0;
            health -= attackMap.get(i);
            if(health <=0) return -1;
            continue;
        }
        if(health < originHealth){
            health += healing;
            if(continuityBandage == bonusTime){
                health += bonusHealing;
                continuityBandage = 0;
            }
        }
        if(health >= originHealth) health = originHealth;
    }
    answer = health;
    return answer;
}
